File: src/test/java/com/koddy/server/member/domain/repository/query/MentorMainSearchRepositoryFetchAppliedMenteesTest.java
package com.koddy.server.member.domain.repository.query;

import com.koddy.server.coffeechat.domain.model.CoffeeChat;
import com.koddy.server.coffeechat.domain.repository.CoffeeChatRepository;
import com.koddy.server.common.RepositoryTest;
import com.koddy.server.common.fixture.CoffeeChatFixture.MenteeFlow;
import com.koddy.server.common.fixture.MenteeFixture;
import com.koddy.server.member.domain.model.mentee.Mentee;
import com.koddy.server.member.domain.model.mentor.Mentor;
import com.koddy.server.member.domain.repository.MemberRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Import;
import org.springframework.data.domain.Page;

import java.util.Arrays;
import java.util.List;

import static com.koddy.server.common.fixture.CoffeeChatFixture.수요일_1주차_20_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.수요일_1주차_21_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.월요일_1주차_20_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.월요일_1주차_21_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.월요일_2주차_20_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.월요일_2주차_21_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.월요일_3주차_20_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.월요일_3주차_21_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.월요일_4주차_20_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.월요일_4주차_21_00_시작;
import static com.koddy.server.common.fixture.MentorFixture.MENTOR_1;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;

@Import(MentorMainSearchRepositoryImpl.class)
@DisplayName("Member -> MentorMainSearchRepository [fetchAppliedMentees] 테스트")
class MentorMainSearchRepositoryFetchAppliedMenteesTest extends RepositoryTest {
    @Autowired
    private MentorMainSearchRepositoryImpl sut;

    @Autowired
    private MemberRepository memberRepository;

    @Autowired
    private CoffeeChatRepository coffeeChatRepository;

    private Mentor mentor;
    private final Mentee[] mentees = new Mentee[10];

    @BeforeEach
    void setUp() {
        mentor = memberRepository.save(MENTOR_1.toDomain());
        final List<MenteeFixture> fixtures = Arrays.stream(MenteeFixture.values())
                .limit(10)
                .toList();
        Arrays.setAll(mentees, it -> memberRepository.save(fixtures.get(it).toDomain()));
    }

    @Test
    @DisplayName("멘토 자신에게 커피챗을 신청한 멘티를 limit 개수만큼 최근에 신청한 순서대로 조회한다")
    void findAppliedMentees() {
        // given
        final CoffeeChat coffeeChat0 = coffeeChatRepository.save(MenteeFlow.apply(월요일_1주차_20_00_시작, mentees[0], mentor));
        final CoffeeChat coffeeChat1 = coffeeChatRepository.save(MenteeFlow.apply(월요일_1주차_21_00_시작, mentees[1], mentor));
        final CoffeeChat coffeeChat2 = coffeeChatRepository.save(MenteeFlow.apply(월요일_2주차_20_00_시작, mentees[2], mentor));
        final CoffeeChat coffeeChat3 = coffeeChatRepository.save(MenteeFlow.apply(월요일_2주차_21_00_시작, mentees[3], mentor));
        final CoffeeChat coffeeChat4 = coffeeChatRepository.save(MenteeFlow.apply(월요일_3주차_20_00_시작, mentees[4], mentor));
        final CoffeeChat coffeeChat5 = coffeeChatRepository.save(MenteeFlow.apply(월요일_3주차_21_00_시작, mentees[5], mentor));
        final CoffeeChat coffeeChat6 = coffeeChatRepository.save(MenteeFlow.apply(월요일_4주차_20_00_시작, mentees[6], mentor));
        final CoffeeChat coffeeChat7 = coffeeChatRepository.save(MenteeFlow.apply(월요일_4주차_21_00_시작, mentees[7], mentor));
        final CoffeeChat coffeeChat8 = coffeeChatRepository.save(MenteeFlow.apply(수요일_1주차_20_00_시작, mentees[8], mentor));
        final CoffeeChat coffeeChat9 = coffeeChatRepository.save(MenteeFlow.apply(수요일_1주차_21_00_시작, mentees[9], mentor));

        /* limit별 조회 */
        final Page<Mentee> result1 = sut.fetchAppliedMentees(mentor.getId(), 3);
        final Page<Mentee> result2 = sut.fetchAppliedMentees(mentor.getId(), 5);
        final Page<Mentee> result3 = sut.fetchAppliedMentees(mentor.getId(), 7);
        final Page<Mentee> result4 = sut.fetchAppliedMentees(mentor.getId(), 10);

        assertAll(
                () -> assertThat(result1.hasNext()).isTrue(),
                () -> assertThat(result1.getTotalElements()).isEqualTo(10),
                () -> assertThat(result1.getContent()).containsExactly(mentees[9], mentees[8], mentees[7]),
                () -> assertThat(result2.hasNext()).isTrue(),
                () -> assertThat(result2.getTotalElements()).isEqualTo(10),
                () -> assertThat(result2.getContent()).containsExactly(mentees[9], mentees[8], mentees[7], mentees[6], mentees[5]),
                () -> assertThat(result3.hasNext()).isTrue(),
                () -> assertThat(result3.getTotalElements()).isEqualTo(10),
                () -> assertThat(result3.getContent()).containsExactly(
                        mentees[9], mentees[8], mentees[7], mentees[6],
                        mentees[5], mentees[4], mentees[3]
                ),
                () -> assertThat(result4.hasNext()).isFalse(),
                () -> assertThat(result4.getTotalElements()).isEqualTo(10),
                () -> assertThat(result4.getContent()).containsExactly(
                        mentees[9], mentees[8], mentees[7], mentees[6], mentees[5],
                        mentees[4], mentees[3], mentees[2], mentees[1], mentees[0]
                )
        );

        /* cancel 후 limit별 조회 */
        coffeeChat3.cancel();
        coffeeChat5.cancel();
        coffeeChat7.cancel();
        coffeeChat9.cancel();

        final Page<Mentee> result5 = sut.fetchAppliedMentees(mentor.getId(), 3);
        final Page<Mentee> result6 = sut.fetchAppliedMentees(mentor.getId(), 5);
        final Page<Mentee> result7 = sut.fetchAppliedMentees(mentor.getId(), 7);
        final Page<Mentee> result8 = sut.fetchAppliedMentees(mentor.getId(), 10);

        assertAll(
                () -> assertThat(result5.hasNext()).isTrue(),
                () -> assertThat(result5.getTotalElements()).isEqualTo(6),
                () -> assertThat(result5.getContent()).containsExactly(mentees[8], mentees[6], mentees[4]),
                () -> assertThat(result6.hasNext()).isTrue(),
                () -> assertThat(result6.getTotalElements()).isEqualTo(6),
                () -> assertThat(result6.getContent()).containsExactly(mentees[8], mentees[6], mentees[4], mentees[2], mentees[1]),
                () -> assertThat(result7.hasNext()).isFalse(),
                () -> assertThat(result7.getTotalElements()).isEqualTo(6),
                () -> assertThat(result7.getContent()).containsExactly(mentees[8], mentees[6], mentees[4], mentees[2], mentees[1], mentees[0]),
                () -> assertThat(result8.hasNext()).isFalse(),
                () -> assertThat(result8.getTotalElements()).isEqualTo(6),
                () -> assertThat(result8.getContent()).containsExactly(mentees[8], mentees[6], mentees[4], mentees[2], mentees[1], mentees[0])
        );
    }
}

File: src/test/kotlin/com/koddy/server/member/domain/repository/query/MentorMainSearchRepositoryFetchAppliedMenteesTest.kt
package com.koddy.server.member.domain.repository.query

import com.koddy.server.coffeechat.domain.model.CoffeeChat
import com.koddy.server.coffeechat.domain.model.CoffeeChatStatus.CANCEL_FROM_MENTEE_FLOW
import com.koddy.server.coffeechat.domain.repository.CoffeeChatRepository
import com.koddy.server.common.RepositoryTestKt
import com.koddy.server.common.fixture.CoffeeChatFixture.수요일_1주차_20_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.수요일_1주차_21_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.월요일_1주차_20_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.월요일_1주차_21_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.월요일_2주차_20_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.월요일_2주차_21_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.월요일_3주차_20_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.월요일_3주차_21_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.월요일_4주차_20_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.월요일_4주차_21_00_시작
import com.koddy.server.common.fixture.MenteeFixtureStore.menteeFixture
import com.koddy.server.common.fixture.MenteeFlow
import com.koddy.server.common.fixture.MentorFixtureStore.mentorFixture
import com.koddy.server.member.domain.model.mentee.Mentee
import com.koddy.server.member.domain.model.mentor.Mentor
import com.koddy.server.member.domain.repository.MemberRepository
import com.koddy.server.member.domain.repository.query.response.AppliedCoffeeChatsByMentee
import io.kotest.assertions.assertSoftly
import io.kotest.matchers.collections.shouldContainExactly
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.context.annotation.Import
import org.springframework.data.domain.Page

@RepositoryTestKt
@Import(MentorMainSearchRepositoryImpl::class)
@DisplayName("Member -> MentorMainSearchRepository [fetchAppliedCoffeeChatsByMentee] 테스트")
internal class MentorMainSearchRepositoryFetchAppliedMenteesTest(
    private val sut: MentorMainSearchRepositoryImpl,
    private val memberRepository: MemberRepository,
    private val coffeeChatRepository: CoffeeChatRepository,
) {
    private lateinit var mentor: Mentor
    private lateinit var mentees: List<Mentee>

    @BeforeEach
    fun setUp() {
        mentor = memberRepository.save(mentorFixture(sequence = 1).toDomain())
        mentees = mutableListOf<Mentee>().apply {
            (1..10).forEach { add(memberRepository.save(menteeFixture(sequence = it).toDomain())) }
        }
    }

    @Test
    fun `멘토 자신에게 커피챗을 신청한 멘티를 limit 개수만큼 최근에 신청한 순서대로 조회한다`() {
        // given
        val coffeeChats: List<CoffeeChat> = coffeeChatRepository.saveAll(
            listOf(
                MenteeFlow.apply(fixture = 월요일_1주차_20_00_시작, mentee = mentees[0], mentor = mentor),
                MenteeFlow.apply(fixture = 월요일_1주차_21_00_시작, mentee = mentees[1], mentor = mentor),
                MenteeFlow.apply(fixture = 월요일_2주차_20_00_시작, mentee = mentees[2], mentor = mentor),
                MenteeFlow.apply(fixture = 월요일_2주차_21_00_시작, mentee = mentees[3], mentor = mentor),
                MenteeFlow.apply(fixture = 월요일_3주차_20_00_시작, mentee = mentees[4], mentor = mentor),
                MenteeFlow.apply(fixture = 월요일_3주차_21_00_시작, mentee = mentees[5], mentor = mentor),
                MenteeFlow.apply(fixture = 월요일_4주차_20_00_시작, mentee = mentees[6], mentor = mentor),
                MenteeFlow.apply(fixture = 월요일_4주차_21_00_시작, mentee = mentees[7], mentor = mentor),
                MenteeFlow.apply(fixture = 수요일_1주차_20_00_시작, mentee = mentees[8], mentor = mentor),
                MenteeFlow.apply(fixture = 수요일_1주차_21_00_시작, mentee = mentees[9], mentor = mentor),
            ),
        )

        /* limit별 조회 */
        val result1: Page<AppliedCoffeeChatsByMentee> = sut.fetchAppliedMentees(mentorId = mentor.id, limit = 3)
        val result2: Page<AppliedCoffeeChatsByMentee> = sut.fetchAppliedMentees(mentorId = mentor.id, limit = 5)
        val result3: Page<AppliedCoffeeChatsByMentee> = sut.fetchAppliedMentees(mentorId = mentor.id, limit = 7)
        val result4: Page<AppliedCoffeeChatsByMentee> = sut.fetchAppliedMentees(mentorId = mentor.id, limit = 10)

        assertSoftly {
            result1.hasNext() shouldBe true
            result1.totalElements shouldBe 10
            result1.content.map { it.coffeeChatId } shouldContainExactly listOf(coffeeChats[9].id, coffeeChats[8].id, coffeeChats[7].id)
            result1.content.map { it.menteeId } shouldContainExactly listOf(mentees[9].id, mentees[8].id, mentees[7].id)
            result2.hasNext() shouldBe true
            result2.totalElements shouldBe 10
            result2.content.map { it.coffeeChatId } shouldContainExactly listOf(
                coffeeChats[9].id, coffeeChats[8].id, coffeeChats[7].id,
                coffeeChats[6].id, coffeeChats[5].id,
            )
            result2.content.map { it.menteeId } shouldContainExactly listOf(
                mentees[9].id, mentees[8].id, mentees[7].id,
                mentees[6].id, mentees[5].id,
            )
            result3.hasNext() shouldBe true
            result3.totalElements shouldBe 10
            result3.content.map { it.coffeeChatId } shouldContainExactly listOf(
                coffeeChats[9].id, coffeeChats[8].id, coffeeChats[7].id, coffeeChats[6].id,
                coffeeChats[5].id, coffeeChats[4].id, coffeeChats[3].id,
            )
            result3.content.map { it.menteeId } shouldContainExactly listOf(
                mentees[9].id, mentees[8].id, mentees[7].id, mentees[6].id,
                mentees[5].id, mentees[4].id, mentees[3].id,
            )
            result4.hasNext() shouldBe false
            result4.totalElements shouldBe 10
            result4.content.map { it.coffeeChatId } shouldContainExactly listOf(
                coffeeChats[9].id, coffeeChats[8].id, coffeeChats[7].id, coffeeChats[6].id, coffeeChats[5].id,
                coffeeChats[4].id, coffeeChats[3].id, coffeeChats[2].id, coffeeChats[1].id, coffeeChats[0].id,
            )
            result4.content.map { it.menteeId } shouldContainExactly listOf(
                mentees[9].id, mentees[8].id, mentees[7].id, mentees[6].id, mentees[5].id,
                mentees[4].id, mentees[3].id, mentees[2].id, mentees[1].id, mentees[0].id,
            )
        }

        /* cancel 후 limit별 조회 */
        coffeeChats[3].cancel(status = CANCEL_FROM_MENTEE_FLOW, cancelBy = mentees[3].id, cancelReason = "취소..")
        coffeeChats[5].cancel(status = CANCEL_FROM_MENTEE_FLOW, cancelBy = mentees[5].id, cancelReason = "취소..")
        coffeeChats[7].cancel(status = CANCEL_FROM_MENTEE_FLOW, cancelBy = mentees[7].id, cancelReason = "취소..")
        coffeeChats[9].cancel(status = CANCEL_FROM_MENTEE_FLOW, cancelBy = mentees[9].id, cancelReason = "취소..")

        val result5: Page<AppliedCoffeeChatsByMentee> = sut.fetchAppliedMentees(mentorId = mentor.id, limit = 3)
        val result6: Page<AppliedCoffeeChatsByMentee> = sut.fetchAppliedMentees(mentorId = mentor.id, limit = 5)
        val result7: Page<AppliedCoffeeChatsByMentee> = sut.fetchAppliedMentees(mentorId = mentor.id, limit = 7)
        val result8: Page<AppliedCoffeeChatsByMentee> = sut.fetchAppliedMentees(mentorId = mentor.id, limit = 10)

        assertSoftly {
            result5.hasNext() shouldBe true
            result5.totalElements shouldBe 6
            result5.content.map { it.coffeeChatId } shouldContainExactly listOf(coffeeChats[8].id, coffeeChats[6].id, coffeeChats[4].id)
            result5.content.map { it.menteeId } shouldContainExactly listOf(mentees[8].id, mentees[6].id, mentees[4].id)
            result6.hasNext() shouldBe true
            result6.totalElements shouldBe 6
            result6.content.map { it.coffeeChatId } shouldContainExactly listOf(
                coffeeChats[8].id, coffeeChats[6].id, coffeeChats[4].id,
                coffeeChats[2].id, coffeeChats[1].id,
            )
            result6.content.map { it.menteeId } shouldContainExactly listOf(
                mentees[8].id, mentees[6].id, mentees[4].id,
                mentees[2].id, mentees[1].id,
            )
            result7.hasNext() shouldBe false
            result7.totalElements shouldBe 6
            result7.content.map { it.coffeeChatId } shouldContainExactly listOf(
                coffeeChats[8].id, coffeeChats[6].id, coffeeChats[4].id,
                coffeeChats[2].id, coffeeChats[1].id, coffeeChats[0].id,
            )
            result7.content.map { it.menteeId } shouldContainExactly listOf(
                mentees[8].id, mentees[6].id, mentees[4].id,
                mentees[2].id, mentees[1].id, mentees[0].id,
            )
            result8.hasNext() shouldBe false
            result8.totalElements shouldBe 6
            result8.content.map { it.coffeeChatId } shouldContainExactly listOf(
                coffeeChats[8].id, coffeeChats[6].id, coffeeChats[4].id,
                coffeeChats[2].id, coffeeChats[1].id, coffeeChats[0].id,
            )
            result8.content.map { it.menteeId } shouldContainExactly listOf(
                mentees[8].id, mentees[6].id, mentees[4].id,
                mentees[2].id, mentees[1].id, mentees[0].id,
            )
        }
    }
}
